### ISSUE TITLE
GIL Not Released When doing Blocking I/O

### ISSUE THREAD

**Issue by ddasilva** [Unidata/netcdf4-python#369]
I was looking at the source, and I noticed the GIL is not released when doing blocking I/O. For those of us that get performance boosts from parallel threads doing I/O, this would block the benefit. Am I mistaken that the GIL is still being held?

There are a pair of macros for releasing and re-acquiring the GIL: `Py_BEGIN_ALLOW_THREADS` and `Py_END_ALLOW_THREADS`.
https://docs.python.org/2/c-api/init.html#releasing-the-gil-from-extension-code

**Comment by jswhit** [comment]
You are correct, the GIL is not released by the Cython extension code before calling nc_put_vara/nc_put_vars.  Whether this would be a good idea or not, I am not sure .

**Comment by jswhit** [comment]
The netcdf library is apparently not thread-safe, so I don't think releasing the GIL would have any benefit.

https://www.unidata.ucar.edu/software/netcdf/docs/faq.html#notthreadsafe

**Comment by shoyer** [comment]
Releasing the GIL is definitely a good idea when possible. See, for example: http://matthewrocklin.com/blog/work/2015/03/10/PyData-GIL/

I actually looked into this a few months ago, but didn't couldn't obtain any benefit in my experiments for reads over a single thread. I thought I was probably doing something wrong, but I'm guessing now that the netCDF C library may have some sort of global lock of its own.

**Comment by jswhit** [comment]
Here's some info on parallelism and the netcdf lib:

http://www.unidata.ucar.edu/support/help/MailArchives/netcdf/msg09903.html

I guess I'm not understanding how releasing the GIL would help unless shared-memory parallelism can be exploited within the netcdf and hdf5 libs.  Do you have a certain use-case in mind?

**Comment by dopplershift** [comment]
My guess would be that if we released the GIL, then other Python code could concurrently do other, non-netcdf stuff within the interpreter. While the GIL is held, that's not possible. You wouldn't get any benefits for free (i.e. without using the threading module), but while the GIL is held it's impossible to see any benefits.

**Comment by shoyer** [comment]
Supporting even just parallel reads would couple very nicely with deferred computation systems such as Blaze/dask: http://matthewrocklin.com/blog/work/2015/02/13/Towards-OOC-Slicing-and-Stacking/

These systems can execute tasks concurrently using multiple threads for major speed ups. For most IO bound tasks multithreading works much better than multiprocessing.

**Comment by shoyer** [comment]
This will couple especially nicely with opendap bindings, too, though I'm not sure if the netcdf library allows multiple threads there. This would particularly be the case for accessing a small amount of data from many variables, which can be really slow if you don't access the data concurrently (and most of that time is just spend waiting on the network).

**Comment by ddasilva** [comment]
@jswhit
>  Do you have a certain use-case in mind?

I am working with remote sensing data where every timestep matrix is in a separate file. I want to be able to read multiple files concurrently. On my system, I have gotten everywhere from 5X - 15X speedups by using the multiprocessing module to read and process the files concurrently, communicating scalar results through pipes. This works well for simple `map()`-like computations, but when the computation is a function of multiple timesteps, the above approach requires sending large matrices between processes.

I/O will be the vast majority of the work, even if I manage to read files concurrently across threads. I'm willing to use threads and throw away the ability to do the computation concurrently, as long as I can do the I/O concurrently.

@dopplershift 
> Granted, the original commenter mentions parallel I/O, which I agree, releasing the GIL would do nothing to help when using standard (non-parallel) NetCDF. Maybe parallel I/O across different files....?

Exactly.

**Comment by jswhit** [comment]
OK, I understand now.  Releasing the GIL can allow threads to process different files concurrently in a more efficient way.   Can we come up with a simple test case to look at the performance benefits of releasing the GIL?

**Comment by dopplershift** [comment]
Maybe this? I haven't tried yet with the GIL released.

```python
from netCDF4 import Dataset
import numpy as np
import threading
import Queue
import time

# Make some files
for i in range(4):
    nc = Dataset('test%d.nc' % i, 'w')
    data = np.random.randn(500, 500, 500)
    nc.createDimension('x', 500)
    nc.createDimension('y', 500)
    nc.createDimension('z', 500)
    var = nc.createVariable('grid', 'f', ('x', 'y', 'z'))
    var[:] = data
    nc.close()

# Queue them up    
items = Queue.Queue()
for i in range(4):
    items.put('test%d.nc' % i)

# Function for threads to use
def get_data():
    fname = items.get()
    print 'got', fname
    nc = Dataset(fname, 'r')
    data = nc.variables['grid'][:]
    nc.close()
    items.task_done()
    print 'finished', fname

# Time it    
start = time.time()
for i in range(4):
    threading.Thread(target=get_data).start()
items.join()
end = time.time()
print end - start
```

**Comment by shoyer** [comment]
I usually use `from multiprocessing.dummy import Pool as ThreadPool`. Then you can use `pool = ThreadPool(N)` and see if the time it takes to run `pool.map(get_data, files)` changes with `N`.

**Comment by jswhit** [comment]
Running @dopplershift's script as is takes 1.12 seconds on my mac.

If I then wrap the calls to nc_get_vara and nc_get_vars with with nogil, the execution time is 0.58 seconds. 

Here's the diff:

``diff --git a/netCDF4.pxi b/netCDF4.pxi
index 20a7345..d7515dd 100644
--- a/netCDF4.pxi
+++ b/netCDF4.pxi
@@ -301,17 +301,17 @@ cdef extern from "netcdf.h":
     int nc_put_var1(int ncid, int varid, size_t *indexp,
                 void *op)
     int nc_get_var1(int ncid, int varid,  size_t *indexp,
-                void *ip)
+                void *ip)
     int nc_put_vara(int ncid, int varid,  size_t *startp,
                 size_t *countp, void *op)
     int nc_get_vara(int ncid, int varid,  size_t *startp,
-                size_t *countp, void *ip)
+                size_t *countp, void *ip) nogil
     int nc_put_vars(int ncid, int varid,  size_t *startp,
                 size_t *countp, ptrdiff_t *stridep,
                 void *op)
     int nc_get_vars(int ncid, int varid,  size_t *startp,
                 size_t *countp, ptrdiff_t *stridep,
-                void *ip)
+                void *ip) nogil
     int nc_put_varm(int ncid, int varid,  size_t *startp,
                 size_t *countp, ptrdiff_t *stridep,
                 ptrdiff_t *imapp, void *op)

diff --git a/netCDF4.pyx b/netCDF4.pyx
index ac3e285..9c036fe 100644
--- a/netCDF4.pyx
+++ b/netCDF4.pyx
@@ -3708,11 +3708,13 @@ The default value of C{mask} is C{True}
             data = numpy.empty(shapeout, self.dtype)
             # strides all 1 or scalar variable, use get_vara (faster)
             if sum(stride) == ndims or ndims == 0:
-                ierr = nc_get_vara(self._grpid, self._varid,
-                                   startp, countp, data.data)
+                with nogil:
+                    ierr = nc_get_vara(self._grpid, self._varid,
+                                       startp, countp, data.data)
             else:
-                ierr = nc_get_vars(self._grpid, self._varid,
-                                   startp, countp, stridep, data.data)
+                with nogil:
+                    ierr = nc_get_vars(self._grpid, self._varid,
+                                       startp, countp, stridep, data.data)
             if ierr == NC_EINVALCOORDS:
                 raise IndexError
             elif ierr != NC_NOERR:``

**Comment by ddasilva** [comment]
@jswhit I take it that was tested on a consumer-level hard drive? RAID setups give larger benefits than 2X when the file is distributed among multiple disks. The file-system in my use case is such.

See the comments on concurrency at the bottom of the **RAID 0** section at: 
http://en.wikipedia.org/wiki/RAID#Standard_levels

**Comment by ddasilva** [comment]
I'll give @dopplershift's test a run on my system

**Comment by dopplershift** [comment]
Personally, I think @jswhit 's test already demonstrated enough that we should be doing this. I can't think of a reason not to. At least, I thought the issue was only whether or not it would make a noticable difference, not necessarily how well it scales.

**Comment by ddasilva** [comment]
If @jswhit test sufficient I will stop here. I would need to wait until tomorrow anyway since my system is missing some compile dependencies and I don't have root.

**Comment by jswhit** [comment]
Questions:

* Should we just wrap every nc_* call with "with nogil:"?  
* Or just the ones involved with read-only access? 
* Since the netcdf lib is apparently not thread-safe, are there any potential pitfalls?

**Comment by jswhit** [comment]
Some information on thread safety

* in HDF5:  http://www.hdfgroup.uiuc.edu/papers/features/mthdf/
* in netcdf: I *think* nc_create(), nc_open(), and nc_close()  are the only functions where thread-safety is an issue.

**Comment by lesserwhirls** [comment]
We build a threads safe version of netCDF-C for use in the THREDDS Data Server, and the document @jswhit points to is all we do differently. Basically, build HDF with:

``
./configure ... --enable-threadsafe --with-pthread=<DIR>
``

where **DIR** is the location of the posix thread library.

*So far, so good* (in terms of writing out netCDF-4 files from the TDS).

**Comment by ddasilva** [comment]
@lesserwhirls I assume that's writing different files concurrently, yes? It might be wise to stop people from trying to write to different parts of the same file concurrently (which would require a parallel netcdf library).

**Comment by lesserwhirls** [comment]
@ddasilva - yes. Here is a common use case for the TDS: using the NetcdfSubset service, users make a request for subsetting (spatially/temporally/variable space) a dataset (could be a variety of data formats on disk) and requests that the server return a netCDF-4 file. Each user gets their own netCDF-4 file fulfilling the request in return.

**Comment by jswhit** [comment]
Created a branch ``nogi`` which wraps performance critical nc_* calls ``with nogil``.  For starters nc_get_var and nc_put_var are wrapped.  Please try this with real-world code and report back here.

**Comment by dopplershift** [comment]
@shoyer Thanks for pointing out ThreadPool, multithreading is not something I do often in Python.

**Comment by shoyer** [comment]
I ran the `nogil` branch through our xray test suite. Everything looks good: https://travis-ci.org/shoyer/xray/jobs/54126532

**Comment by jswhit** [comment]
Thanks @shoyer.  @ddasilva, do you have some threaded code you could test it with? I want to know if it improves performance, and most importantly, gives the right answer.

**Comment by ddasilva** [comment]
@jswhit I put in a request to have your branch installed on my system. I'll get back to you.

**Comment by jswhit** [comment]
I'm still a little nervous that releasing the GIL could cause hard to debug errors in some cases, if you're not using a thread-safe version of the HDF5 library (which hardly anyone is).  I noticed that h5py has a rather elaborate locking mechanism to make sure nothing bad happens in threaded code.

**Comment by shoyer** [comment]
Indeed, we're not (yet!) trying to do any sort of concurrent writes with xray/netCDF4. Still, you have to pretty explicitly choose to do multithreading (in most circumstances).

**Comment by jswhit** [comment]
Do we know for sure that concurrent reads are OK, and there are no global variables in the library that will get trashed?  Seems unlikely, but I'd like to be sure.

**Comment by ddasilva** [comment]
I ran a test on my system, and the nogil branch was much more efficient. The below plot shows a comparison with master.

> Do we know for sure that concurrent reads are OK, and there are no global variables in the library 
> that will get trashed? Seems unlikely, but I'd like to be sure.

I too am concerned whether NetCDF library uses one file descriptor per dataset. This would result in inaccurate data if one thread calls `fseek()` while the other is reading chunks.

According to Russ Rew in 2011, concurrent reads to the same file were supported in at least NetCDF-3 classic: http://www.unidata.ucar.edu/support/help/MailArchives/netcdf/msg09903.html

![comparison](https://cloud.githubusercontent.com/assets/536378/6643271/66e3a16a-c981-11e4-8b9f-03b8cfbb0d78.png)

**Comment by dopplershift** [comment]
Just talked to the netCDF maintainer. Concurrent *reads*, even on the same file, should be fine.

**Comment by jswhit** [comment]
@ddasilva, thanks for doing those tests.  Quite impressive speed gains using the nogil branch.  Did you program test for correctness?

Here's an interesting passage from the HDF5 FAQ (http://www.hdfgroup.org/hdf5-quest.html#gconc):

``Users are often surprised to learn that (1) concurrent access to different datasets in a single HDF5 file and (2) concurrent access to different HDF5 files both require a thread-safe version of the HDF5 library. Although each thread in these examples is accessing different data, the HDF5 library modifies global data structures that are independent of a particular HDF5 dataset or HDF5 file. HDF5 relies on a semaphore around the library API calls in the thread-safe version of the library to protect the data structure from corruption by simultaneous manipulation from different threads. Examples of HDF5 library global data structures that must be protected are the freespace manager and open file lists. ``

Not quite sure what all of the means, but it is a bit disturbing.  I don't know if the netcdf layer has a locking mechanism to prevent this from happening.  So far none of my tests have found any problems with concurrent reads of the same file or different files with threads with the GIL released.

**Comment by jswhit** [comment]
If there are no objections, I'll merge this tomorrow.  I put an entry in the Changelog that emphasizes that recompilation of HDF5 (using --with-threadsafe) may be necessary to prevent global data corruption when reading from multiple threads.

**Comment by jswhit** [comment]
merged pull request #370

**Comment by shoyer** [comment]
@ddasilva are your files netCDF3 or netCDF4? I've been trying to figure out the story for concurrency with HDF5 recently (see https://github.com/h5py/h5py/issues/591)... we've had a least one report of seg faults when using netCDF4 with multiple threads, so I've actually added a thread lock in xray. Might be something we want to use in netCDF4-python, too.

**Comment by ddasilva** [comment]
@shoyer My test was with NetCDF4. According to @jswhit 's quote from the HDF5 doc (above), concurrent access in HDF5 is not supported.
